File: src/Lungs.cpp
#include "MedicalLib/Lungs.h"
#include "MedicalLib/Patient.h" // For Blood struct
#include <random>
#include <algorithm>
#include <sstream>
#include <cmath>
#include <numeric>

// Helper for random fluctuations
static double getFluctuation(double stddev) {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    std::normal_distribution<> d(0, stddev);
    return d(gen);
}

Lungs::Lungs(int id)
    : Organ(id, "Lungs"),
      respirationRate(16.0),
      oxygenSaturation(98.0),
      tidalVolume_mL(500.0),
      endTidalCO2_mmHg(40.0),
      peakInspiratoryPressure_cmH2O(0.0),
      totalLungCapacity_mL(6000.0),
      currentState(RespiratoryState::PAUSE),
      cyclePosition_s(0.0),
      totalTime_s(0.0),
      capnographyHistorySize(200) {

    // Initialize Lobes
    rightUpperLobe = {"Right Upper Lobe", 0, 0.1};
    rightMiddleLobe = {"Right Middle Lobe", 0, 0.07};
    rightLowerLobe = {"Right Lower Lobe", 0, 0.13};
    leftUpperLobe = {"Left Upper Lobe", 0, 0.1};
    leftLowerLobe = {"Left Lower Lobe", 0, 0.1};

    // Initialize Bronchi
    mainBronchus = {"Main Bronchus", 0.8};
}

void Lungs::update(Patient& patient, double deltaTime_s) {
    totalTime_s += deltaTime_s;

    updateRespiratoryMechanics(deltaTime_s);
    updateGasLevels(deltaTime_s);

    // Update capnography data
    capnographyData.push_front(generateCapnographyValue());
    if (capnographyData.size() > capnographyHistorySize) {
        capnographyData.pop_back();
    }

    // --- Blood Interaction ---
    Blood& blood = patient.blood;
    // Gas exchange is driven by the pressure gradient between alveoli and blood
    double ventilationFactor = (tidalVolume_mL / 500.0) * (respirationRate / 16.0);
    ventilationFactor = std::clamp(ventilationFactor, 0.5, 1.5); // Clamp effect

    // Oxygenation: Blood O2 moves towards the lung's O2 level
    double o2_gradient = oxygenSaturation - blood.oxygenSaturation;
    blood.oxygenSaturation += o2_gradient * 0.8 * ventilationFactor * deltaTime_s;
	blood.oxygenSaturation = std::clamp(blood.oxygenSaturation, 0.0, 100.0);

    // CO2 Removal: Blood CO2 moves towards the lung's (low) CO2 level
    // We'll model the lung's CO2 as being lower than the blood's target
    double effectiveAlveolarCO2 = 40.0 / ventilationFactor;
    double co2_gradient = blood.co2PartialPressure_mmHg - effectiveAlveolarCO2;
    blood.co2PartialPressure_mmHg -= co2_gradient * 0.5 * deltaTime_s;
	blood.co2PartialPressure_mmHg = std::clamp(blood.co2PartialPressure_mmHg, 0.0, 200.0);
}

void Lungs::updateRespiratoryMechanics(double deltaTime_s) {
    double cycleDuration_s = 60.0 / respirationRate;
    cyclePosition_s += deltaTime_s;

    // State transitions
    double inspirationDuration = cycleDuration_s * 0.4; // I:E ratio of 1:1.5
    double expirationDuration = cycleDuration_s * 0.6;

    if (cyclePosition_s <= inspirationDuration) {
        currentState = RespiratoryState::INSPIRATION;
    } else if (cyclePosition_s <= cycleDuration_s) {
        currentState = RespiratoryState::EXPIRATION;
    } else {
        cyclePosition_s -= cycleDuration_s;
        currentState = RespiratoryState::INSPIRATION;
    }

    // Pressure and Volume dynamics
    double totalCompliance = rightUpperLobe.compliance + rightMiddleLobe.compliance + rightLowerLobe.compliance + leftUpperLobe.compliance + leftLowerLobe.compliance;

    double flowRate_mL_s = 0;
    if (currentState == RespiratoryState::INSPIRATION) {
        // Simple sine wave for pressure generation
        double pressure_wave = sin(M_PI * (cyclePosition_s / inspirationDuration));
        peakInspiratoryPressure_cmH2O = 15.0 * pressure_wave; // 15 cmH2O peak
        flowRate_mL_s = (peakInspiratoryPressure_cmH2O / mainBronchus.resistance) * 100 * totalCompliance;
        tidalVolume_mL += flowRate_mL_s * deltaTime_s;
    } else { // EXPIRATION
        peakInspiratoryPressure_cmH2O = 0;
        // Passive recoil drives expiration
        double recoilPressure = (tidalVolume_mL / 500.0) * 5.0; // Simplified
        flowRate_mL_s = -(recoilPressure / mainBronchus.resistance) * 100;
        tidalVolume_mL += flowRate_mL_s * deltaTime_s;
    }

    // Clamp tidal volume
    tidalVolume_mL = std::clamp(tidalVolume_mL, 0.0, totalLungCapacity_mL / 2.0);
}

void Lungs::updateGasLevels(double deltaTime_s) {
    // SpO2 is affected by how well we are breathing
    double ventilationFactor = (tidalVolume_mL / 500.0) * (respirationRate / 16.0);
    double targetSpo2 = 98.0 * std::clamp(ventilationFactor, 0.9, 1.0);
    oxygenSaturation += 0.1 * (targetSpo2 - oxygenSaturation) * deltaTime_s + getFluctuation(0.02);
    oxygenSaturation = std::clamp(oxygenSaturation, 94.0, 100.0);

    // etCO2 is inversely related to ventilation
    double targetEtCO2 = 40.0 / std::clamp(ventilationFactor, 0.8, 1.2);
    endTidalCO2_mmHg += 0.2 * (targetEtCO2 - endTidalCO2_mmHg) * deltaTime_s + getFluctuation(0.05);
    endTidalCO2_mmHg = std::clamp(endTidalCO2_mmHg, 35.0, 50.0);
}

double Lungs::generateCapnographyValue() {
    double cycleDuration_s = 60.0 / respirationRate;
    double timeInCycle = fmod(cyclePosition_s, cycleDuration_s);
    double inspirationEnd = cycleDuration_s * 0.4;
    double plateauStart = cycleDuration_s * 0.5;
    double plateauEnd = cycleDuration_s * 0.8;

    if (currentState == RespiratoryState::INSPIRATION) {
        return 0.0; // Phase I: Inspiratory baseline
    } else { // EXPIRATION
        if (timeInCycle < plateauStart) { // Phase II: Expiratory upstroke
            return endTidalCO2_mmHg * ((timeInCycle - inspirationEnd) / (plateauStart - inspirationEnd));
        } else if (timeInCycle < plateauEnd) { // Phase III: Alveolar plateau
            return endTidalCO2_mmHg + getFluctuation(0.1);
        } else { // Phase IV: Inspiratory downstroke (handled by next cycle's baseline)
            return endTidalCO2_mmHg * (1.0 - (timeInCycle - plateauEnd) / (cycleDuration_s - plateauEnd));
        }
    }
}

std::string Lungs::getSummary() const {
    std::stringstream ss;
    ss.precision(1);
    ss << std::fixed;
    ss << "--- Lungs Summary ---\n"
       << "Respiration Rate: " << getRespirationRate() << " breaths/min\n"
       << "Oxygen Saturation (SpO2): " << getOxygenSaturation() << " %\n"
       << "Tidal Volume: " << getTidalVolume() << " mL\n"
       << "End-Tidal CO2 (etCO2): " << getEndTidalCO2() << " mmHg\n"
       << "Peak Airway Pressure: " << getPeakInspiratoryPressure() << " cmH2O\n";
    return ss.str();
}

// --- Getters Implementation ---
void Lungs::inflictDamage(double damage) {
    double damageFactor = 1.0 - std::clamp(damage, 0.0, 1.0);
    rightUpperLobe.compliance *= damageFactor;
    rightMiddleLobe.compliance *= damageFactor;
    rightLowerLobe.compliance *= damageFactor;
    leftUpperLobe.compliance *= damageFactor;
    leftLowerLobe.compliance *= damageFactor;
}

double Lungs::getRespirationRate() const { return respirationRate; }
double Lungs::getOxygenSaturation() const { return oxygenSaturation; }
double Lungs::getTidalVolume() const { return tidalVolume_mL; }
double Lungs::getEndTidalCO2() const { return endTidalCO2_mmHg; }
double Lungs::getPeakInspiratoryPressure() const { return peakInspiratoryPressure_cmH2O; }
const std::deque<double>& Lungs::getCapnographyWaveform() const { return capnographyData; }

// --- Setters Implementation ---
void Lungs::setRespirationRate(double newRate_bpm) {
    respirationRate = newRate_bpm;
}

File: src/organs/lungs.rs
//! Lungs organ simulation
//!
//! Simulates respiratory mechanics and gas exchange including:
//! - 5 lobes (right upper/middle/lower, left upper/lower)
//! - Respiratory cycle (inspiration, expiration, pause)
//! - Oxygen saturation
//! - CO2 exchange and capnography

use crate::organ::{Organ, OrganId};
use crate::patient::Patient;
use std::collections::VecDeque;

/// Respiratory phase
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum RespiratoryPhase {
    Inspiration,
    Expiration,
    Pause,
}

/// Lung lobe
#[derive(Debug, Clone)]
pub struct Lobe {
    pub name: String,
    pub volume_ml: f64,
    pub compliance: f64,  // 0.0 = no compliance, 1.0 = normal
    pub ventilation_rate: f64,
}

/// Bronchus
#[derive(Debug, Clone)]
pub struct Bronchus {
    pub name: String,
    pub resistance: f64,  // Airway resistance
}

/// Lungs organ
#[derive(Debug)]
pub struct Lungs {
    id: OrganId,
    /// Right upper lobe
    pub right_upper_lobe: Lobe,
    /// Right middle lobe
    pub right_middle_lobe: Lobe,
    /// Right lower lobe
    pub right_lower_lobe: Lobe,
    /// Left upper lobe
    pub left_upper_lobe: Lobe,
    /// Left lower lobe
    pub left_lower_lobe: Lobe,
    /// Main bronchus
    pub main_bronchus: Bronchus,
    /// Respiratory rate (breaths per minute)
    pub respiration_rate_bpm: f64,
    /// Tidal volume (mL)
    pub tidal_volume_ml: f64,
    /// Oxygen saturation (%)
    pub oxygen_saturation_percent: f64,
    /// End-tidal CO2 (mmHg)
    pub end_tidal_co2_mmhg: f64,
    /// Peak inspiratory pressure (cmH2O)
    pub peak_inspiratory_pressure: f64,
    /// Capnography waveform
    pub capnography_waveform: VecDeque<f64>,
    /// Current respiratory phase
    pub current_phase: RespiratoryPhase,
    /// Internal respiratory cycle timer
    respiratory_cycle_time: f64,
}

impl Lungs {
    /// Create new lungs
    pub fn new(id: OrganId) -> Self {
        Self {
            id,
            right_upper_lobe: Lobe {
                name: "Right Upper".to_string(),
                volume_ml: 600.0,
                compliance: 1.0,
                ventilation_rate: 1.0,
            },
            right_middle_lobe: Lobe {
                name: "Right Middle".to_string(),
                volume_ml: 500.0,
                compliance: 1.0,
                ventilation_rate: 1.0,
            },
            right_lower_lobe: Lobe {
                name: "Right Lower".to_string(),
                volume_ml: 800.0,
                compliance: 1.0,
                ventilation_rate: 1.0,
            },
            left_upper_lobe: Lobe {
                name: "Left Upper".to_string(),
                volume_ml: 600.0,
                compliance: 1.0,
                ventilation_rate: 1.0,
            },
            left_lower_lobe: Lobe {
                name: "Left Lower".to_string(),
                volume_ml: 800.0,
                compliance: 1.0,
                ventilation_rate: 1.0,
            },
            main_bronchus: Bronchus {
                name: "Main".to_string(),
                resistance: 1.0,
            },
            respiration_rate_bpm: 16.0,
            tidal_volume_ml: 500.0,
            oxygen_saturation_percent: 98.0,
            end_tidal_co2_mmhg: 38.0,
            peak_inspiratory_pressure: 15.0,
            capnography_waveform: VecDeque::with_capacity(1000),
            current_phase: RespiratoryPhase::Pause,
            respiratory_cycle_time: 0.0,
        }
    }

    /// Inflict damage to a lobe
    pub fn inflict_damage(&mut self, lobe_index: usize, damage_amount: f64) {
        let lobe = match lobe_index {
            0 => &mut self.right_upper_lobe,
            1 => &mut self.right_middle_lobe,
            2 => &mut self.right_lower_lobe,
            3 => &mut self.left_upper_lobe,
            4 => &mut self.left_lower_lobe,
            _ => return,
        };
        lobe.compliance = (lobe.compliance - damage_amount).max(0.0);
    }

    /// Calculate total lung compliance
    fn total_compliance(&self) -> f64 {
        (self.right_upper_lobe.compliance
            + self.right_middle_lobe.compliance
            + self.right_lower_lobe.compliance
            + self.left_upper_lobe.compliance
            + self.left_lower_lobe.compliance)
            / 5.0
    }
}

impl Organ for Lungs {
    fn update(&mut self, patient: &mut Patient, delta_time_s: f64) {
        // Update respiratory cycle
        self.respiratory_cycle_time += delta_time_s;
        let cycle_duration = 60.0 / self.respiration_rate_bpm;

        if self.respiratory_cycle_time >= cycle_duration {
            self.respiratory_cycle_time = 0.0;
        }

        let cycle_progress = self.respiratory_cycle_time / cycle_duration;

        // Determine respiratory phase
        if cycle_progress < 0.4 {
            self.current_phase = RespiratoryPhase::Inspiration;
        } else if cycle_progress < 0.8 {
            self.current_phase = RespiratoryPhase::Expiration;
        } else {
            self.current_phase = RespiratoryPhase::Pause;
        }

        // Calculate gas exchange based on compliance
        let compliance_factor = self.total_compliance();
        let effective_ventilation = self.tidal_volume_ml * compliance_factor;

        // Oxygen saturation
        self.oxygen_saturation_percent = 98.0 * compliance_factor;
        self.oxygen_saturation_percent = self.oxygen_saturation_percent.clamp(70.0, 100.0);

        // CO2 clearance
        let _co2_clearance = effective_ventilation * 0.05;
        self.end_tidal_co2_mmhg = 38.0 + (1.0 - compliance_factor) * 20.0;

        // Update patient blood gases
        patient.blood.gases.sao2_percent = self.oxygen_saturation_percent;
        patient.blood.gases.paco2_mmhg = self.end_tidal_co2_mmhg;
        patient.blood.gases.pao2_mmhg = self.oxygen_saturation_percent * 0.95; // Approximate PaO2 from SpO2

        // Peak inspiratory pressure affected by compliance
        self.peak_inspiratory_pressure = 15.0 / compliance_factor.max(0.1);

        // Generate capnography waveform
        let capno_value = match self.current_phase {
            RespiratoryPhase::Inspiration => 0.0,
            RespiratoryPhase::Expiration => {
                if cycle_progress < 0.6 {
                    self.end_tidal_co2_mmhg * (cycle_progress - 0.4) / 0.2
                } else {
                    self.end_tidal_co2_mmhg
                }
            }
            RespiratoryPhase::Pause => 0.0,
        };

        self.capnography_waveform.push_back(capno_value);
        if self.capnography_waveform.len() > 1000 {
            self.capnography_waveform.pop_front();
        }

        // Respond to blood chemistry
        // High CO2 increases respiration rate
        if patient.blood.gases.paco2_mmhg > 45.0 {
            self.respiration_rate_bpm = 16.0 + (patient.blood.gases.paco2_mmhg - 45.0) * 0.5;
            self.respiration_rate_bpm = self.respiration_rate_bpm.min(30.0);
        } else {
            self.respiration_rate_bpm = 16.0;
        }
    }

    fn get_summary(&self) -> String {
        format!(
            "Lungs: RR={:.0} bpm, TV={:.0} mL, SpO2={:.1}%, etCO2={:.1} mmHg, PIP={:.1} cmH2O",
            self.respiration_rate_bpm,
            self.tidal_volume_ml,
            self.oxygen_saturation_percent,
            self.end_tidal_co2_mmhg,
            self.peak_inspiratory_pressure
        )
    }

    fn get_id(&self) -> OrganId {
        self.id
    }

    fn get_type(&self) -> &'static str {
        "Lungs"
    }

    fn as_any(&self) -> &dyn std::any::Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
        self
    }
}
